src/d3d9/d3d9_vertex_declaration.cpp

```diff
@@ -205,6 +206,106 @@ namespace dxvk {
+  // This table maps members of a D3DVERTEXELEMENT9 declaration to a FVF code.
+  //
+  // Data type              Usage                       Usage index     FVF
+  // ----------------------------------------------------------------------------------------
+  // D3DDECLTYPE_FLOAT3     D3DDECLUSAGE_POSITION       0               D3DFVF_XYZ
+  // D3DDECLTYPE_FLOAT4     D3DDECLUSAGE_POSITIONT      0               D3DFVF_XYZRHW
+  // D3DDECLTYPE_FLOATn     D3DDECLUSAGE_BLENDWEIGHT    0               D3DFVF_XYZBn
+  // D3DDECLTYPE_UBYTE4     D3DDECLUSAGE_BLENDINDICES   0               D3DFVF_XYZB(nWeights + 1)
+  // D3DDECLTYPE_FLOAT3     D3DDECLUSAGE_NORMAL         0               D3DFVF_NORMAL
+  // D3DDECLTYPE_FLOAT1     D3DDECLUSAGE_PSIZE          0               D3DFVF_PSIZE
+  // D3DDECLTYPE_D3DCOLOR   D3DDECLUSAGE_COLOR          0               D3DFVF_DIFFUSE
+  // D3DDECLTYPE_D3DCOLOR   D3DDECLUSAGE_COLOR          1               D3DFVF_SPECULAR
+  // D3DDECLTYPE_FLOATm     D3DDECLUSAGE_TEXCOORD       n               D3DFVF_TEXCOORDSIZEm(n)
+  // D3DDECLTYPE_FLOAT3     D3DDECLUSAGE_POSITION       1               N / A
+  // D3DDECLTYPE_FLOAT3     D3DDECLUSAGE_NORMAL         1               N / A
+
+  if (element.Usage == D3DDECLUSAGE_POSITION && element.Type == D3DDECLTYPE_FLOAT3 && element.UsageIndex == 0)
+    return D3DFVF_XYZ;
+  else if (element.Usage == D3DDECLUSAGE_POSITIONT && element.Type == D3DDECLTYPE_FLOAT4 && element.UsageIndex == 0)
+    return D3DFVF_XYZRHW;
+  else if (element.Usage == D3DDECLUSAGE_BLENDWEIGHT && element.UsageIndex == 0) {
+    switch (element.Type)
+    {
+    case D3DDECLTYPE_FLOAT1:
+      return D3DFVF_XYZB1;
+    case D3DDECLTYPE_FLOAT2:
+      return D3DFVF_XYZB2;
+    case D3DDECLTYPE_FLOAT3:
+      return D3DFVF_XYZB3;
+    case D3DDECLTYPE_FLOAT4:
+      return D3DFVF_XYZB4;
+    default:;
+    }
+  }
+  else if (element.Usage == D3DDECLUSAGE_BLENDINDICES && element.Type == D3DDECLTYPE_UBYTE4 && element.UsageIndex == 0)
+    return D3DFVF_XYZB1;
+  else if (element.Usage == D3DDECLUSAGE_NORMAL && element.Type == D3DDECLTYPE_FLOAT3 && element.UsageIndex == 0)
+    return D3DFVF_NORMAL;
+  else if (element.Usage == D3DDECLUSAGE_PSIZE && element.Type == D3DDECLTYPE_FLOAT1 && element.UsageIndex == 0)
+    return D3DFVF_PSIZE;
+  else if (element.Usage == D3DDECLUSAGE_COLOR && element.Type == D3DDECLTYPE_D3DCOLOR) {
+    switch (element.UsageIndex)
+    {
+    case 0:
+      return D3DFVF_DIFFUSE;
+    case 1:
+      return D3DFVF_SPECULAR;
+    default:;
+    }
+  }
+  else if (element.Usage == D3DDECLUSAGE_TEXCOORD && element.UsageIndex < 8) {
+      DWORD localFvf = 0;
+      do {
+        // Check if bits of format for current UsageIndex are free in the fvf
+        // It is necessary to skip multiple initializations of the bitfield because
+        // returned value is bitwise or-ed to final fvf DWORD.
+        if (D3DFVF_TEXCOORDSIZE1(element.UsageIndex) & fvf != 0)
```

Comment:
Make this a while loop with this as the condition? :P

Reply:
moved to the helper